<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.10: http://docutils.sourceforge.net/" />
<title>SangerMutantLibraryAnalysis</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7514 2012-09-14 14:27:12Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="sangermutantlibraryanalysis">
<h1 class="title">SangerMutantLibraryAnalysis</h1>

<div class="section" id="summary">
<h1>Summary</h1>
<p>This is a script for analyzing the distribution of mutations among Sanger-sequenced clones in a mutant library of a protein-coding gene. The distributions of mutations are analyzed and plotted.</p>
<p>The script for this analysis is available <a class="reference external" href="https://github.com/jbloom/SangerMutantLibraryAnalysis">on GitHub</a>.</p>
<p>This script was written by <a class="reference external" href="http://research.fhcrc.org/bloom/en.html">Jesse Bloom</a>.</p>
</div>
<div class="section" id="analyses">
<h1>Analyses</h1>
<p>Here are the analyses (each in its own subdirectory) performed with this script:</p>
<ul class="simple">
<li><tt class="docutils literal">2013Analysis_Influenza_NP_Aichi681</tt> contains the results of analyzing a mutant library of the A/Aichi/2/1968 (H3N2) influenza gene. Analysis performed by <a class="reference external" href="http://research.fhcrc.org/bloom/en.html">Jesse Bloom</a>.</li>
</ul>
</div>
<div class="section" id="requirements">
<h1>Requirements</h1>
<p>This analysis simply consists of a <a class="reference external" href="http://www.python.org/">Python</a> script. It has been tested with <a class="reference external" href="http://www.python.org/">Python</a> versions 2.6 and 2.7, and probably works with other 2.* versions as well.</p>
<p>The script requires <a class="reference external" href="http://www.scipy.org/">scipy</a> and <a class="reference external" href="http://matplotlib.org/">matplotlib</a>. It has been tested with <a class="reference external" href="http://www.scipy.org/">scipy</a> 0.12.0 and <a class="reference external" href="http://matplotlib.org/">matplotlib</a> version 1.2.1, but will probably work with other versions as well.</p>
<p>The script uses <a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a> to convert <tt class="docutils literal">*.pdf</tt> files to <tt class="docutils literal">*.jpg</tt> files.</p>
</div>
<div class="section" id="running-the-script">
<h1>Running the script</h1>
<p>The analysis is performed by the script <em>analyze_library.py</em>. To run the script, simply go to the current directory and type the command:</p>
<pre class="literal-block">
python analyze_library.py
</pre>
<p>The script will then ask you to enter the names of two input files: the sequence file, and the mutation list file. These are both text files that should have the following format:</p>
<blockquote>
<ul>
<li><p class="first"><strong>The sequence file:</strong> this is simply a FASTA file that contains a single protein-coding gene. This should be the gene that you are sequencing. For example, here is an example of such a file (<tt class="docutils literal"><span class="pre">Aichi68-NP.fasta</span></tt>):</p>
<pre class="literal-block">
&gt;Aichi68-NP
atggcgtcccaaggcaccaaacggtcttatgaacagatggaaactgatggggaacgccagaatgcaactgagatcagagcatccgtcgggaagatgattgatggaattggacgattctacatccaaatgtgcactgaacttaaactcagtgattatgaggggcgactgatccagaacagcttaacaatagagagaatggtgctctctgcttttgacgaaagaaggaataaatatctggaagaacatcccagcgcggggaaggatcctaagaaaactggaggacccatatacaagagagtagatagaaagtggatgagggaactcgtcctttatgacaaagaagaaataaggcgaatctggcgccaagccaataatggtgatgatgcaacagctggtctgactcacatgatgatctggcattccaatttgaatgatacaacataccagaggacaagagctcttgttcgcaccggcatggatcccaggatgtgctctctgatgcagggttcgactctccctaggaggtctggagctgcaggcgctgcagtcaaaggagttgggacaatggtgatggagttgataaggatgatcaaacgtgggatcaatgatcggaacttctggagaggtgaaaatggacgaaaaacaaggagtgcttacgagagaatgtgcaacattctcaaaggaaaatttcaaacagctgcacaaagggcaatgatggatcaagtgagagaaagtcggaacccaggaaatgctgagatcgaagatctcatctttctggcacggtctgcactcatattgagagggtcagttgctcacaaatcttgtctgcccgcctgtgtgtatggacctgccgtagccagtggctacgacttcgaaaaagagggatactctttagtgggaatagaccctttcaaactgcttcaaaacagccaagtatacagcctaatcagaccgaacgagaatccagcacacaagagtcagctggtgtggatggcatgcaattctgctgcatttgaagatctaagagtattaagcttcatcagagggaccaaagtatccccaagggggaaactttccactagaggagtacaaattgcttcaaatgaaaacatggatgctatggaatcaagtactcttgaactgagaagcaggtactgggccataagaaccagaagtggaggaaacactaatcaacagagggcctctgcaggtcaaatcagtgtgcaacctgcattttctgtgcaaagaaacctcccatttgacaaaccaaccatcatggcagcattcactgggaatacagagggaagaacatcagacatgagggcagaaattataaggatgatggaaggtgcaaaaccagaagaaatgtccttccaggggcggggagtcttcgagctctcggacgaaagggcagcgaacccgatcgtgccctcttttgacatgagtaatgaaggatcttatttcttcggagacaatgcagaggagtacgacaattaa
</pre>
</li>
<li><p class="first"><strong>The mutation list file:</strong> this is a text file that lists the mutations. The mutations should be numbered in sequential (1, 2, ...) numbering according to the sequence specified in the sequence file. Lines in this file that are empty or begin with the character # are ignored. All other lines should specify a clone and all identified mutations. The clone name should be the first entry on the line, followed by a colon. There is then a comma-delimited list of the mutations. The mutations are indicated as follows:</p>
<blockquote>
<ul class="simple">
<li>Single nucleotide substitutions are indicated as <em>G1378T</em> for mutation of site 1378 from <em>G</em> to <em>T</em>.</li>
<li>Multiple-nucleotide mutations at the same codon should be listed as <em>AG349GA</em> or <em>TCT1003GCC</em>. List mutations like this if they are sequential (or two mutations separated by a single other mutation) as these are probably mutations of the same codon.</li>
<li>Deletions should be listed like this: <em>delC392</em> for deletion of the <em>C</em> at position 392.</li>
<li>Insertions should be listed like this: <em>insG392</em> for insertion of a <em>G</em> at position 392.</li>
<li>If a clone has not mutations, you should enter <em>None</em>.</li>
</ul>
</blockquote>
<p>The script will check that the specified wildtype nucleotides actually match thos e indicated in the sequence file. If they do not, an error will be raised. Note that even if your sequence contains an insertion or deletion, you must ensure that subsequent sites are still numbered according to sequential numbering of the provided sequence.</p>
<p>Here is an example input file (<tt class="docutils literal">mutation_list.txt</tt>):</p>
<pre class="literal-block">
# Sequential 1, 2, ... numbering from start codon of Aichi68-NP coding sequence
#
# Sequencing from Jan-25-2013 for clones 1 to 3 for each.
# Sequencing from Jan-26-2013 for clones 4 to 8 for each.
WT1round3-1: delCT392, AG583CT
WT1round3-2: ATG991TCC, AA1249CG, delA1257, TT1390GA
WT1round3-3: AAA271CGG, TT1285AC, G1378T
WT1round3-4: AA175GC, CTC1396GAA
WT1round3-5: TAT232GTA, ACA388TAC, CT1007AC, AAT1492GCA
WT1round3-6: AAC925GGG, GGA1303TCG
WT1round3-7: None
WT1round3-8: GGA553ATC, A696C, TG713AT, CA1208GT
#
WT2round3-1: GC1148AA
WT2round3-2: AC16CT, AG269GC, TTC1390AGA
WT2round3-3: CCC481AAG, AC1168GT
WT2round3-4: TTT211GTA, C322G
WT2round3-6: AAA817TTT, AGA1306CTC, delA1311
WT2round3-7: AG92CC, ACA436GTC, ATG1441GAA
WT2round3-8: C397T, TA854CT, CCG952ACA, A1314T, GAG1483CCC
#
N334H1round3-1: GT808AC
N334H1round3-2: None
N334H1round3-3: GAT379TAA, G662T, CG1415GC
N334H1round3-4: A1078G
N334H1round3-5: T825C, AT1111TA, AAT1189GCG, ACA1309CCT
N334H1round3-6: G241A, AAA550TGG
N334H1round3-7: TCT1237CCG, TC1286GG
N334H1round3-8: TG130AC, AG349GA, G391C, AT487TA, AG799CA
#
N334H2round3-1: ATG313CGT, GC929CA, GTG985TGA, AAT1492TTG
N334H2round3-2: A12C, A1171G, CCA1267GTG
N334H2round3-3: AGA1024CTG
N334H2round3-4: G6C, GG160TT, CC265TT, GGT376AAC, C512T, TCT1003GCC
N334H2round3-6: CCC247TTT, TCT784GAG
N334H2round3-7: ACT67GTG, AGA193CGT, T616A, TAC1153GTA, AGT1174CAC
N334H2round3-8: AT425GC, CC470AT, AGA1165GTC, GCA1324CAG
</pre>
</li>
</ul>
</blockquote>
</div>
<div class="section" id="output-of-the-script">
<h1>Output of the script</h1>
<p>The script will print some information about the mutation statistics to standard output. It will also create some PDF plot files. For example, running the script using the example sequence file <em>Aichi68-NP.fasta</em> and the example mutation list <em>mutation_list.txt</em> provided with this script <a class="reference external" href="https://github.com/jbloom/SangerMutantLibraryAnalysis">on GitHub</a> will produce the following information printed to standard output:</p>
<pre class="literal-block">
Beginning analysis.

Enter the name of the FASTA file containing the gene sequence: Aichi68-NP.fasta
Read a coding sequence of length 1497

Enter the name of the file containing the list of mutations: mutation_list.txt

Reading mutations from mutation_list.txt
Read entries for 30 clones

Substitutions begin at following positions: 2, 4, 6, 23, 31, 44, 54, 59, 65, 71, 78, 81, 83, 89, 90, 91, 105, 108, 117, 126, 127, 130, 131, 133, 142, 146, 157, 161, 163, 171, 184, 185, 195, 206, 221, 232, 238, 262, 267, 270, 273, 275, 285, 309, 310, 318, 329, 331, 335, 336, 342, 360, 371, 383, 385, 389, 390, 391, 392, 397, 403, 413, 417, 423, 429, 429, 435, 436, 437, 438, 442, 460, 464, 464, 466, 472, 481, 495, 498, 498

Indels begin at following positions: 131, 419, 437

Found a total of 80 substitutions out of 14970 codons sequenced (0.0053)

Here are the fractions with different numbers of nucleotide mutations:
  1 nucleotide mutations: 0.00100
  2 nucleotide mutations: 0.00227
  3 nucleotide mutations: 0.00207

Here are the fractions of mutation types
  synonymous: 0.00040
  nonsynonymous: 0.00494
  stop codon: 0.00000

 Now creating the output PDF plot files...
 The output PDF file plots have now all been created.

 Script complete.
</pre>
<p>The produced PDF files are as follows:</p>
<div class="figure align-center">
<img alt="mutpositions.pdf" src="2013Analysis_Influenza_NP_Aichi68/mutpositions.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">mutpositions.pdf</tt> uses lines to show the locations of point mutations and insertions and deletions along the primary sequence.</p>
</div>
<div class="figure align-center">
<img alt="mutpositions_cumulative.pdf" src="2013Analysis_Influenza_NP_Aichi68/mutpositions_cumulative.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">mutpositions_cumulative.pdf</tt> shows the cumulative distribution of substitution mutations along the primary sequence versus the line expected for a uniform distribution of mutations.</p>
</div>
<div class="figure align-center">
<img alt="ncodonmuts.pdf" src="2013Analysis_Influenza_NP_Aichi68/ncodonmuts.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">ncodonmuts.pdf</tt> shows the average number of nucleotide changes for each codon substitution mutation versus the expectation if each codon is equally likely to be mutated to every other codon.</p>
</div>
<div class="figure align-center">
<img alt="nmutdist.pdf" src="2013Analysis_Influenza_NP_Aichi68/nmutdist.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">nmutdist.pdf</tt> shows the distribution of the number of codon substitution mutations per clone versus the expectation for a Poisson distribution.</p>
</div>
<div class="figure align-center">
<img alt="codonmutntcomposition.pdf" src="2013Analysis_Influenza_NP_Aichi68/codonmutntcomposition.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">codonmutntcomposition.pdf</tt> shows the nucleotide composition (fraction of each of the nucleotides) in the parent and mutant codons for each mutation. If your library is totally unbiased for every codon mutating to every other codon with equal probability, the parent composition should be equal to the nucleotide composition of your gene (which is probably not exactly 25% of each nucleotide), whereas the mutant composition should be very close to 25% of each nucleotide.</p>
</div>
<div class="figure align-center">
<img alt="mutationclustering.pdf" src="2013Analysis_Influenza_NP_Aichi68/mutationclustering.jpg" style="width: 35%;" />
<p class="caption">The above plot <tt class="docutils literal">mutationclustering.pdf</tt> helps you determine if mutations are clustered in primary sequence in clones with multiple substitutions. This plot is less intuitive than the others. For every clone with multiple mutations, the pairwise distance between those mutations in primary sequence is computed. The actual cumulative distribution of these pairwise distances is shown. The script then simulates sets of clones that have the exact same number of mutations per clone as observed in the actual mutation list with these mutations located randomly in primary sequence. It then computes the cumulative distribution of the pairwise distances for 1000 of these simulated libraries, and plots that. If the two distributions look similar, you library probably does not have a tendency for clustering of mutations in primary sequence. If the distributions are different, such clustering may exist.</p>
</div>
</div>
</div>
</body>
</html>
